File: Source/LemmatizerLib/MorphAutomat.cpp
// ==========  This file is under  LGPL, the GNU Lesser General Public Licence
// ==========  Dialing Lemmatizer (www.aot.ru), 
// ==========  Copyright by Alexey Sokirko (2004)

#include "StdMorph.h"
#include "MorphAutomat.h"
#include "../common/bserialize.h"

#ifdef DETECT_MEMORY_LEAK
	#ifdef _DEBUG
	#define new DEBUG_NEW

	#undef THIS_FILE
	static char THIS_FILE[] = __FILE__;
	#endif
#endif


static int  InitAlphabet(MorphLanguageEnum Language, int* pCode2Alphabet, int *pAlphabet2Code, size_t AnnotChar)
{
	assert (!is_upper_alpha(AnnotChar, Language));
	string AdditionalEnglishChars = "'1234567890";
	string AdditionalGermanChars = "";
	int AlphabetSize = 0;
	for (size_t i=0; i < 256; i++)
	{
		if	(		is_upper_alpha((BYTE)i, Language) 
				||	(i == '-') 
				||	(i == AnnotChar)
				||	(		(Language == morphEnglish) 
						&&	(AdditionalEnglishChars.find(i) != string::npos)
					)
				||	(		(Language == morphGerman) 
						&&	(AdditionalGermanChars.find(i) != string::npos)
					)
				||	(		(Language == morphURL) 
					  &&	is_alpha(i, morphURL)
					 )
			)
		{
			pCode2Alphabet[AlphabetSize] = i;
			pAlphabet2Code[i] = AlphabetSize;
			AlphabetSize++;
		}
		else
			pAlphabet2Code[i] = -1;

	};

	if (AlphabetSize > MaxAlphabetSize)
	{
		string Error = "Error! The  ABC is too large";
		ErrorMessage (Error);
		throw CExpc(Error);
	};

	return AlphabetSize;
};
string CABCEncoder::GetCriticalNounLetterPack() const
{
	return string(MinimalPredictionSuffix, m_AnnotChar);
}

string CABCEncoder::EncodeIntToAlphabet(DWORD v) const
{
	string Result;
	if (v == 0)
	{
		Result.push_back(m_Code2AlphabetWithoutAnnotator[0]);
		const char* debug = Result.c_str();
		return Result;
	}
	else
	while (v > 0)
	{
		Result.push_back(m_Code2AlphabetWithoutAnnotator[v%m_AlphabetSizeWithoutAnnotator]);
		v  /= m_AlphabetSizeWithoutAnnotator;
	};
	return Result;
};

DWORD CABCEncoder::DecodeFromAlphabet(const string& v) const
{
	size_t len = v.length();
	int c = 1;
	int Result = 0;
	for (size_t i=0; i <len; i++)
	{
		Result += m_Alphabet2CodeWithoutAnnotator[(BYTE)v[i]]*c;
		c*=m_AlphabetSizeWithoutAnnotator;
	};
	return Result;
};

bool  CABCEncoder::CheckABCWithAnnotator(const string& WordForm) const
{
	size_t Length = WordForm.length();
	for (size_t i=0; i < Length; i++)
		if (m_Alphabet2Code[(BYTE)WordForm[i]] == -1)
			return false;
	return true;
};
bool  CABCEncoder::CheckABCWithoutAnnotator(const string& WordForm) const
{
	size_t Length = WordForm.length();
	for (size_t i=0; i < Length; i++)
		if (m_Alphabet2CodeWithoutAnnotator[(BYTE)WordForm[i]] == -1)
			return false;
	return true;
};


CABCEncoder::CABCEncoder(MorphLanguageEnum Language, BYTE AnnotChar) : m_AnnotChar(AnnotChar)
{
	m_AlphabetSize = ::InitAlphabet(Language,m_Code2Alphabet,m_Alphabet2Code,m_AnnotChar);
	m_AlphabetSizeWithoutAnnotator = ::InitAlphabet(Language,m_Code2AlphabetWithoutAnnotator,m_Alphabet2CodeWithoutAnnotator,257/* non-exeting symbol */);
	assert (m_AlphabetSizeWithoutAnnotator + 1 == m_AlphabetSize);
	m_Language = Language;
};



CMorphAutomat::CMorphAutomat(MorphLanguageEnum Language, BYTE AnnotChar) : CABCEncoder(Language, AnnotChar)
{
	m_pNodes = 0;
	m_pRelations = 0;
	m_NodesCount = 0;
	m_RelationsCount = 0;
};

CMorphAutomat::~CMorphAutomat()
{
	Clear();
};

void CMorphAutomat::Clear()
{
	if (m_pNodes)
		delete m_pNodes;
	m_pNodes = 0;
	m_NodesCount = 0;

	if (m_pRelations)
		delete m_pRelations;
	m_pRelations = 0;
	m_RelationsCount = 0;
};


inline size_t get_size_in_bytes (const CMorphAutomNode& t)
{
	return		
			get_size_in_bytes(t.m_Data)
			;
};

inline size_t save_to_bytes(const CMorphAutomNode& t, BYTE* buf)
{
	buf += save_to_bytes(t.m_Data, buf);
	return get_size_in_bytes(t);
};

inline size_t restore_from_bytes(CMorphAutomNode& t, const BYTE* buf)
{
	buf += restore_from_bytes(t.m_Data, buf);
	return get_size_in_bytes(t);
};



inline size_t get_size_in_bytes (const CMorphAutomRelation& t)
{
	return		get_size_in_bytes(t.m_Data);
};

inline size_t save_to_bytes(const CMorphAutomRelation& t, BYTE* buf)
{
	buf += save_to_bytes(t.m_Data, buf);
	return get_size_in_bytes(t);
};

inline size_t restore_from_bytes(CMorphAutomRelation& t, const BYTE* buf)
{
	buf += restore_from_bytes(t.m_Data, buf);
	return get_size_in_bytes(t);
};





void CMorphAutomat::BuildChildrenCache()
{
	size_t Count = ChildrenCacheSize;
	if (m_NodesCount < ChildrenCacheSize)
		Count = m_NodesCount;

	m_ChildrenCache.resize(Count*MaxAlphabetSize, -1);
	
	
	for (size_t NodeNo=0; NodeNo <Count; NodeNo++)
	{
		const CMorphAutomRelation* start = m_pRelations+m_pNodes[NodeNo].GetChildrenStart();
		const CMorphAutomRelation* end = start + GetChildrenCount(NodeNo);
		for (; start != end; start++)
		{
			const CMorphAutomRelation& p = *start;
			m_ChildrenCache[NodeNo*MaxAlphabetSize+m_Alphabet2Code[p.GetRelationalChar()]] = p.GetChildNo();
		};
	};
	
};




bool CMorphAutomat::Load(string AutomatFileName)
{
	Clear();
	
	FILE * fp = fopen(AutomatFileName.c_str(), "rb");
	if (!fp)
	{
		ErrorMessage (Format("Cannot open %s", AutomatFileName.c_str()));
		return false;
	};
	
	char buffer [256];
	if (!fgets(buffer, 256, fp)) return false;
	m_NodesCount = atoi(buffer);
	if (!m_NodesCount) return false;

	assert (m_pNodes == 0);
	
	m_pNodes  = new CMorphAutomNode[m_NodesCount];
	assert (m_pNodes != 0);
	if (fread(m_pNodes, sizeof(CMorphAutomNode),m_NodesCount, fp) != m_NodesCount)
		return  false;
	

	if (!fgets(buffer, 256, fp)) return false;
	m_RelationsCount = atoi(buffer);
	assert (m_pRelations == 0);
	m_pRelations  = new CMorphAutomRelation[m_RelationsCount];
	assert (m_pRelations != 0);
	if (fread(m_pRelations, sizeof(CMorphAutomRelation),m_RelationsCount, fp) != m_RelationsCount)
		return  false;
	

	{
		int Alphabet2Code[256];
		fread(Alphabet2Code,sizeof(int),256,fp);
		if (memcmp(Alphabet2Code,m_Alphabet2Code, 256*sizeof(int)) )
		{
			string err = Format("%s alphabet has changed; cannot load morph automat", GetStringByLanguage(m_Language).c_str());
			ErrorMessage(err);
			return false;
		};
	};


	fclose(fp);


	
	BuildChildrenCache();

	return true;
};

bool CMorphAutomat::Save(string AutomatFileName) const
{
	try {
		FILE * fp = fopen(AutomatFileName.c_str(), "wb");
		if (!fp)
		{
			ErrorMessage (Format("CMorphAutomat::Save, cannot write to %s", AutomatFileName.c_str()));
			return false;
		};
		fprintf(fp, "%i\n", m_NodesCount);
		if (fwrite(m_pNodes, sizeof(CMorphAutomNode),m_NodesCount, fp) != m_NodesCount)
			return  false;

		fprintf(fp, "%i\n", m_RelationsCount);
		if (fwrite(m_pRelations, sizeof(CMorphAutomRelation),m_RelationsCount, fp) != m_RelationsCount)
			return  false;
		

		fwrite(m_Alphabet2Code,sizeof(int),256,fp);

		fclose(fp);

		printf ("%i children\n",m_RelationsCount);
		printf ("%i nodes\n",m_NodesCount);


		return true;
	}
	catch (...)
	{
		fprintf (stderr, "Cannot save CMorphAutomat");
		return false;
	};
};





size_t  CMorphAutomat::GetChildrenCount(size_t NodeNo)  const
{  
	//return m_Nodes[NodeNo].m_ChildrenEndIndex - m_Nodes[NodeNo].m_ChildrenStartIndex;
	if (NodeNo+1 == m_NodesCount)
			return m_RelationsCount - m_pNodes[NodeNo].GetChildrenStart();
	else
			return m_pNodes[NodeNo+1].GetChildrenStart() - m_pNodes[NodeNo].GetChildrenStart();

};


const CMorphAutomRelation*  CMorphAutomat::GetChildren(size_t NodeNo) const 
{ 
	return m_pRelations + m_pNodes[NodeNo].GetChildrenStart(); 
};



void CMorphAutomat::DumpAllStringsRecursive(FILE* fp, int NodeNo, string CurrPath) const 
{
	if (m_pNodes[NodeNo].IsFinal())
	{
		fprintf (fp, "%s\n", CurrPath.c_str());
	};
	size_t Count =  GetChildrenCount(NodeNo);
	for (size_t i=0; i<Count; i++)
	{
		const CMorphAutomRelation& p = GetChildren(NodeNo)[i];
		string q = CurrPath;
		q.push_back(p.GetRelationalChar());
		DumpAllStringsRecursive(fp, p.GetChildNo(), q);
	};

};

bool CMorphAutomat::DumpAllStrings(string FileName) const 
{
	FILE *fp  = fopen (FileName.c_str(), "w");
	if (!fp) return false;
	if (m_NodesCount > 0 ) 
		DumpAllStringsRecursive(fp, 0, "");
	fclose (fp);
	return true;
};



int CMorphAutomat::NextNode(int NodeNo, BYTE RelationChar) const
{
	if (NodeNo < ChildrenCacheSize)
	{
		int z = m_Alphabet2Code[RelationChar];
		if (z == -1) return -1;
		return m_ChildrenCache[NodeNo*MaxAlphabetSize+z];
	}
	else
	{
		const CMorphAutomRelation* start = m_pRelations +m_pNodes[NodeNo].GetChildrenStart();
		const CMorphAutomRelation* end = start + GetChildrenCount(NodeNo);

		for (; start != end; start++)
		{
			const CMorphAutomRelation& p = *start;
			if (RelationChar == p.GetRelationalChar())
				return p.GetChildNo();
		};

		return -1;
	};
};


string	CMorphAutomat::GetFirstResult (const string& Text) const 
{
	int NodeNo = FindStringAndPassAnnotChar(Text, 0);
	if ( NodeNo == -1) return "";
	string res = "";
	while (!m_pNodes[NodeNo].IsFinal())
	{
		const CMorphAutomRelation& p = GetChildren(NodeNo)[0];
		res.push_back(p.GetRelationalChar());
		NodeNo = p.GetChildNo();
	};
	return res;
};

void	CMorphAutomat::GetAllMorphInterpsRecursive (int NodeNo, string& curr_path, vector<CAutomAnnotationInner>& Infos) const 
{
	const CMorphAutomNode& N = m_pNodes[NodeNo];
	if (N.IsFinal())
	{
		CAutomAnnotationInner A;
		DWORD i = DecodeFromAlphabet(curr_path);
		size_t ItemNo;
		size_t ModelNo;
		size_t PrefixNo;
		DecodeMorphAutomatInfo(i, ModelNo, ItemNo, PrefixNo);
		A.m_ItemNo = ItemNo;
		A.m_ModelNo =  ModelNo;
		A.m_PrefixNo = PrefixNo;
		Infos.push_back(A);
	};

	size_t Count =  GetChildrenCount(NodeNo);
	size_t CurrPathSize = curr_path.size();
	curr_path.resize(CurrPathSize + 1);
	for (size_t i=0; i<Count; i++)
	{
		const CMorphAutomRelation& p = GetChildren(NodeNo)[i];
		curr_path[CurrPathSize] = p.GetRelationalChar();
		GetAllMorphInterpsRecursive(p.GetChildNo(), curr_path, Infos);
	};
	curr_path.resize(CurrPathSize);
};

int	CMorphAutomat::FindStringAndPassAnnotChar (const string& Text, size_t TextPos) const 
{
	size_t TextLength = Text.length();
	int r = 0;
	for (size_t i=TextPos; i<TextLength; i++)
	{
		int nd =  NextNode(r,(BYTE)Text[i]);
		
		if (nd ==  -1) 
		{
			return -1;		
		}
		r = nd;
	};
	//assert ( r != -1);
	//  passing annotation char

	return NextNode(r,m_AnnotChar);
};

void	CMorphAutomat::GetInnerMorphInfos (const string& Text, size_t TextPos, vector<CAutomAnnotationInner>& Infos) const 
{
	Infos.clear();
	int r = FindStringAndPassAnnotChar(Text, TextPos);
	if ( r == -1) return;

	// getting all interpretations
	string curr_path;
	GetAllMorphInterpsRecursive(r, curr_path, Infos);

	//assert (!Infos.empty());
	//sort(Infos.begin(),Infos.end());
};

DWORD CMorphAutomat::EncodeMorphAutomatInfo (size_t ModelNo, size_t ItemNo, size_t PrefixNo) const 
{
	return			(((DWORD)((WORD)((DWORD)(ModelNo) & 0xffff))) << 18)
				|	(((DWORD)((WORD)((DWORD)(ItemNo) & 0xffff))) << 9)
				|	PrefixNo;

};

void CMorphAutomat::DecodeMorphAutomatInfo (DWORD Info, size_t& ModelNo, size_t& ItemNo, size_t& PrefixNo) const 
{
	ModelNo = Info >>18;
	ItemNo  = (0x3FFFF&Info) >>9;
	PrefixNo = (0x1FF&Info);
};

File: trunk/Source/LemmatizerNET/Implement/MorphAutomat.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using LemmatizerNET.Implement.Agramtab;
using LemmatizerNET.Files;

namespace LemmatizerNET.Implement {
	internal class MorphAutomat : ABCEncoder {
		private MorphAutomNode[] _nodes;
		private MorphAutomRelation[] _relations;
		private int[] _childrenCache;

		private int FindStringAndPassAnnotChar(string text, int pos) {
			var TextLenth = text.Length;
			var r = 0;
			for (var i = pos; i < TextLenth; i++) {
				var nd = NextNode(r, text[i]);
				if (nd == -1) {
					return -1;
				}
				r = nd;
			}
			return NextNode(r, AnnotChar);
		}
		private void GetAllMorphInterpsRecursive(int nodeNo, string currPath, IList<AutomAnnotationInner> infos) {
			if (_nodes[nodeNo].IsFinal) {
				var a = new AutomAnnotationInner();
				int i = DecodeFromAlphabet(currPath);
				int itemNo;
				int modelNo;
				int prefixNo;
				DecodeMorphAutomatInfo(i, out modelNo, out itemNo, out prefixNo);
				a.ItemNo = (short)itemNo;
				a.ModelNo = (short)modelNo;
				a.PrefixNo = (short)prefixNo;
				infos.Add(a);
			}
			var count = GetChildrenCount(nodeNo);
			var currPathSize = currPath.Length;
			var buff = new char[currPathSize + 1];
			currPath.CopyTo(0, buff, 0, currPathSize);
			for (var i = 0; i < count; i++) {
				var p = GetChildren(nodeNo, i);
				buff[currPathSize] = Tools.GetChar(p.RelationalChar);
				GetAllMorphInterpsRecursive(p.ChildNo, new string(buff), infos);
			}
		}
		private void BuildChildrenCache() {
			var count = Constants.ChildrenCacheSize;
			if (_nodes.Length < Constants.ChildrenCacheSize) {
				count = _nodes.Length;
			}
			int len = count * Constants.MaxAlphabetSize;
			_childrenCache = new int[len];
			for (var i = 0; i < len; i++) {
				_childrenCache[i] = -1;
			}

			for (var NodeNo = 0; NodeNo < count; NodeNo++) {
				var count2 = GetChildrenCount(NodeNo);
				var start = _nodes[NodeNo].ChildrenStart;
				for (var i = 0; i < count2; i++) {
					_childrenCache[NodeNo * Constants.MaxAlphabetSize + _alphabet2Code[_relations[i + start].RelationalChar]] = _relations[i + start].ChildNo;
				}
			}
		}
		private void Clear() {
			_nodes = null;
			_relations = null;
		}
		public MorphAutomat(Lemmatizer lemmatizer,InternalMorphLanguage language, char annotChar)
			: base(lemmatizer,language, annotChar) {
		}
		public int NextNode(int nodeNo, char relationChar) {
			if (nodeNo < Constants.ChildrenCacheSize) {
				int z = _alphabet2Code[Tools.GetByte(relationChar)];
				if (z == -1) {
					return -1;
				}
				return _childrenCache[nodeNo * Constants.MaxAlphabetSize + z];
			} else {
				var num = Tools.GetByte(relationChar);
				var start = _nodes[nodeNo].ChildrenStart;
				var count = GetChildrenCount(nodeNo);
				for (int i = 0; i < count; i++) {
					if (num == _relations[i + start].RelationalChar) {
						return _relations[i + start].ChildNo;
					}
				}
			}
			return -1;
		}
		public MorphAutomNode GetNode(int nodeNo) {
			return _nodes[nodeNo];
		}
		public void GetInnerMorphInfos(string text, int textPos, IList<AutomAnnotationInner> infos) {
			infos.Clear();
			var r = FindStringAndPassAnnotChar(text, textPos);
			if (r == -1) {
				return;
			}
			GetAllMorphInterpsRecursive(r, "", infos);
		}
		public int GetChildrenCount(int nodeNo) {
			if (nodeNo + 1 == _nodes.Length) {
				return _relations.Length - _nodes[nodeNo].ChildrenStart;
			} else {
				return _nodes[nodeNo + 1].ChildrenStart - _nodes[nodeNo].ChildrenStart;
			}
		}
		public bool Load(string grammarFileName,FileManager manager) {
			Clear();
			using (var file = manager.GetFile(Lemmatizer.Registry, grammarFileName)) {
				string line = null;
				var reader = new BinaryReader(file, Tools.InternalEncoding);
				line = Tools.ReadLine(reader);
				if (line == null) {
					return false;
				}
				int nodeCount;
				int.TryParse(line, out nodeCount);
				if (nodeCount <= 0) {
					return false;
				}
				if (_nodes != null) {
					throw new MorphException("_nodes != null");
				}
				var len = nodeCount * sizeof(int);
				var buffer = reader.ReadBytes(len);
				if (buffer.Length != len) {
					return false;
				}
				var bufferInt = new int[nodeCount];
				Buffer.BlockCopy(buffer, 0, bufferInt, 0, buffer.Length);
				_nodes = new MorphAutomNode[nodeCount];
				for (var i = 0; i < nodeCount; i++) {
					_nodes[i] = new MorphAutomNode { Data = bufferInt[i] }; ;
				}
				line = Tools.ReadLine(reader);
				if (line == null) {
					return false;
				}
				int.TryParse(line, out nodeCount);
				len = nodeCount * sizeof(int);
				buffer = reader.ReadBytes(len);
				if (buffer.Length != len) {
					return false;
				}
				bufferInt = new int[nodeCount];
				Buffer.BlockCopy(buffer, 0, bufferInt, 0, buffer.Length);
				_relations = new MorphAutomRelation[nodeCount];
				for (int i = 0; i < nodeCount; i++) {
					_relations[i] = new MorphAutomRelation { Data = bufferInt[i] };
				}
				len = Constants.AlphabetSize * sizeof(int);
				buffer = reader.ReadBytes(len);
				if (buffer.Length != len) {
					return false;
				}
				var buf = new int[Constants.AlphabetSize];
				Buffer.BlockCopy(buffer, 0, buf, 0, buffer.Length);
				if (!Tools.ListEquals(buf, _alphabet2Code)) {
					throw new MorphException(Tools.GetStringByLanguage(Language) + "alphabet has changed; cannot load morph automat");
				}
			}
			BuildChildrenCache();
			return true;
		}
		public MorphAutomRelation GetChildren(int nodeNo,int index) {
			return _relations[_nodes[nodeNo].ChildrenStart+index];
		}
		public int EncodeMorphAutomatInfo(int modelNo, int itemNo, int prefixNo) {
				return (int)((((uint)((ushort)((uint)(modelNo) & 0xffff))) << 18)
					| (((uint)((ushort)((uint)(itemNo) & 0xffff))) << 9)
					| prefixNo);
		}
		public void DecodeMorphAutomatInfo(int info, out int modelNo, out int itemNo, out int prefixNo) {
			modelNo = info >> 18;
			itemNo = (0x3FFFF & info) >> 9;
			prefixNo = (0x1FF & info);
		}
		public string GetFirstResult(string text) {
			var nodeNo = FindStringAndPassAnnotChar(text, 0);
			if (nodeNo == -1) {
				return "";
			}
			var res = "";
			while (!_nodes[nodeNo].IsFinal) {
				var p = GetChildren(nodeNo,0);
				res += p.RelationalChar;
				nodeNo = p.ChildNo;
			}
			return res;
		}
		//private void DumpAllStringsRecursive(Stream fp, int NodeNo, string CurrPath);
		//internal bool DumpAllStrings(string FileName);
		//internal bool Save(string GrammarFileName); //TODO Future
	}
}
